File: src/test/java/org/stella/MainTest.java
package org.stella;

import org.junit.jupiter.api.*;
import org.junit.jupiter.params.*;
import org.junit.jupiter.params.provider.*;

import static org.junit.jupiter.api.Assertions.*;
import java.io.*;

class MainTest {


    @ParameterizedTest(name = "{index} Typechecking well-typed program {0}")
    @ValueSource(strings = {
            "tests/well-typed/factorial.stella",
            "tests/well-typed/squares.stella",
            "tests/well-typed/increment_twice.stella",
            "tests/well-typed/logical-operators.stella"})
    public void testWellTyped(String filepath) throws IOException, Exception {
        String[] args = new String[0];
        final InputStream original = System.in;
        final FileInputStream fips = new FileInputStream(new File(filepath));
        System.setIn(fips);
        Main.main(args);
        System.setIn(original);
    }

    @ParameterizedTest(name = "{index} Typechecking ill-typed program {0}")
    @ValueSource(strings = {
            "tests/ill-typed/bad-squares-1.stella",
            "tests/ill-typed/bad-squares-2.stella"})
    public void testIllTyped(String filepath) throws IOException, Exception {
        String[] args = new String[0];
        final InputStream original = System.in;
        final FileInputStream fips = new FileInputStream(new File(filepath));
        System.setIn(fips);
        try {
            Main.main(args);
            throw new Exception("expected the typechecker to fail!");
        } catch (Exception e) {
            // TODO: check that there is a type error actually, and not a problem with implementation
        }
        System.setIn(original);
    }
}

File: src/test/kotlin/org/stella/MainTest.kt
package org.stella

import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource
import java.io.FileInputStream
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.util.stream.Stream

internal class MainTest {
    @ParameterizedTest(name = "{index} Typechecking well-typed program {0}")
    @MethodSource("wellTypedPathStream")
    @Throws(
        IOException::class,
        Exception::class
    )
    fun testWellTyped(filepath: Path) {
        println(filepath.toUri())
        val original = System.`in`
        val fips = FileInputStream(filepath.toFile())
        System.setIn(fips)
        main()
        System.setIn(original)
    }

    @ParameterizedTest(name = "{index} Typechecking ill-typed program {0}")
    @MethodSource("illTypedPathStream")
    @Throws(
        IOException::class,
        Exception::class
    )
    fun testIllTyped(filepath: Path) {
        println(filepath.toUri())
        val original = System.`in`
        val fips = FileInputStream(filepath.toFile())
        System.setIn(fips)
        assertThrows<Exception>("expected the typechecker to fail!") {
            main()
        }
        System.setIn(original)
    }

    companion object {
        private const val BASE_DIR = "src/test/resources"
        private const val WELL_TYPED_DIR = "$BASE_DIR/well-typed"
        private const val ILL_TYPED_DIR = "$BASE_DIR/ill-typed"

        @JvmStatic
        fun wellTypedPathStream(): Stream<Path> = getFilesStream(WELL_TYPED_DIR)

        @JvmStatic
        fun illTypedPathStream(): Stream<Path> = getFilesStream(ILL_TYPED_DIR)

        private fun getFilesStream(path: String) = Files.list(Paths.get(path))
    }
}
